src/util/generate/detect-from-extensions.ts

```diff
@@ -0,0 +1,120 @@
+  let depth = 0
+  let diveDepth = 0
+  let count = 0
+
+  const extensions = map.extensions[extension]
+  if (extensions) {
+    depth++
+    if (typeof extensions === 'number') {
+      count += extensions
+    } else {
+      count++
+    }
+
+    Object.keys(map.dir).forEach((name) => {
+      const dive = getCountAndDepth(extension, map.dir[name])
+      if (diveDepth < dive.depth) diveDepth = dive.depth
+      count += dive.count
+    })
+    depth += diveDepth
+  }
+
+  return { depth, count }
+}
+
+export function getExtensionOptions(extension: string, type: 'single' | 'many', recovery: 'destructure' | 'manual') {
+  const upload = locale.upload[type]
+  const ignore = locale.ignore[type]
+  const options = extensions[extension]
+    ? extensions[extension].map((builder) => ({
+        [builder]: locale[builder]
+          ? locale[builder][type]
+          : `${type === 'single' ? 'This file' : 'These files'} should be built`
+      })).reduce((acc, val) => Object.assign(acc, val, {}))
+    : {}
+
+  return {
+    ...options,
+    upload,
+    ignore,
+    [recovery]: locale[recovery][type]
+  }
+}
+
+export function getManualOptions(type: 'single' | 'many'): { [key: string]: string } {
+  let options: { [key: string]: string } = {}
+  Object.values(builders).forEach((builder) => {
+    options[builder] = locale[builder]
+      ? locale[builder][type]
+      : `${type === 'single' ? 'This file' : 'These files'} should be built by ${builder}`
+  })
+
+  return options
+}
+
+type helpers = { choose: chooseType }
+export async function detectFromExtensions(map: DirMap, deepCapture: string[], rel: string, { choose }: helpers) {
+  const builds: Build[] = []
+  const capture: string[] = []
+
+  for (let ext in map.extensions) {
+    if (Object.prototype.hasOwnProperty.call(map.extensions, ext) && !deepCapture.includes(ext)) {
+      const { depth, count } = getCountAndDepth(ext, map)
+      let use
+      let src
+
+      if (depth > 1) {
+        use = await choose(
+          `What are the ${count} ${ext} files in .${sep}${rel} (${depth} deep)?`,
+          getExtensionOptions(ext, 'many', 'destructure')
+        )
+      }
+
+      if (use && use !== 'destructure') {
+        capture.push(ext)
+        src = `${rel ? `${rel.replace('\\', '/')}/` : ''}**/*${ext}`
+      } else {
+        src = `${rel ? `${rel.replace('\\', '/')}/` : ''}*${ext}`
+        if (typeof map.extensions[ext] === 'string') {
+          const question = `What is .${sep}${join(rel, map.extensions[ext] as string)}?`
+
+          use = await choose(
+            question,
+            getExtensionOptions(ext, 'single', 'manual')
+          )
+
+          if (use === 'manual') {
+            use = await choose(
+              question,
+              getManualOptions('single')
+            )
+          }
+        } else {
+          const question = `What are the ${map.extensions[ext]} ${ext} files in .${sep}${rel} (shallow)?`
```

Comment:
What does `shallow` mean?

Reply:
3 files, shallow (1 deep)
```
/
|- about.js
|- blog.js
|- contact.js
```
3 files, 2 deep
```
/
|- blog/
|  |- index.js
|- about/
|  |- index.js
|- contact.js
```
5 files, 3 deep
```
/
|- blog/
|  |- details/
|  |  |- index.js // 3 deep
|  |  |- update.js // 3 deep
|  |- index.js
|- about/
|  |- index.js // 2 deep
|- contact.js // 1 deep
```

Reply:
Ok that is starting to make sense.
I think this could use some improvement but I'm not sure how to solve it.
Perhaps you should show `*.js` or `**/*.js` when appropriate instead of the count?
cc @leo

Reply:
If it goes too deep, use `destructure`